nimble/host/services/cts/src/ble_svc_cts.c

```diff
@@ -0,0 +1,544 @@
+            0, /* No more characteristics in this service. */
+        } },
+    },
+
+    {
+        0, /* No more services. */
+    },
+};
+
+static inline int
+ble_svc_cts_current_time_read_access(struct ble_gatt_access_ctxt *ctxt)
+{
+    /* Get current clock time */
+    struct os_timeval  tv;
+    struct os_timezone tz;
+    struct clocktime   ct;
+    
+    if (os_gettimeofday(&tv, &tz) || timeval_to_clocktime(&tv, &tz, &ct)) {
+	return BLE_ATT_ERR_UNLIKELY;
+    }
+
+    /* Populate data */
+    ble_svc_cts_current_time.year          = htole16(ct.year);
+    ble_svc_cts_current_time.month         = ct.mon;
+    ble_svc_cts_current_time.day           = ct.day;
+    ble_svc_cts_current_time.hours         = ct.hour;
+    ble_svc_cts_current_time.minutes       = ct.min;
+    ble_svc_cts_current_time.secondes      = ct.sec;
+    ble_svc_cts_current_time.day_of_week   = (ct.dow == 0) ? 7 : ct.dow;
+    ble_svc_cts_current_time.fraction256   = (ct.usec * 256) / 1000000;
+    ble_svc_cts_current_time.adjust_reason = ble_svc_cts_adjust_reason;
+
+    /* Copy data */
+    if (os_mbuf_append(ctxt->om, &ble_svc_cts_current_time,
+		                 sizeof(ble_svc_cts_current_time))) {
+	return BLE_ATT_ERR_INSUFFICIENT_RES;
+    }
+
+    return 0;
+}
+
+static inline int
+ble_svc_cts_current_time_write_access(struct ble_gatt_access_ctxt *ctxt)
+{
+    int rc = 0;
+
+    /* Retrieve data */
+    uint16_t om_len = OS_MBUF_PKTLEN(ctxt->om);
+    if (om_len != sizeof(ble_svc_cts_current_time)) {
+        return BLE_ATT_ERR_INVALID_ATTR_VALUE_LEN;
+    }
+    if (ble_hs_mbuf_to_flat(ctxt->om,
+			    &ble_svc_cts_current_time, om_len, NULL)) {
+        return BLE_ATT_ERR_UNLIKELY;
+    }
+    
+    /* Convert to timeval */
+    struct clocktime   ct;
+    struct os_timezone tz;
+    struct os_timeval  tv;
+
+    if (os_gettimeofday(NULL, &tz)) {
+        return BLE_ATT_ERR_UNLIKELY;
+    }
+    
+    ct.year = le16toh(ble_svc_cts_current_time.year);
+    ct.mon  = ble_svc_cts_current_time.month;
+    ct.day  = ble_svc_cts_current_time.day;
+    ct.hour = ble_svc_cts_current_time.hours;
+    ct.min  = ble_svc_cts_current_time.minutes;
+    ct.sec  = ble_svc_cts_current_time.secondes;
+    ct.usec = (ble_svc_cts_current_time.fraction256 * 1000000) / 256;
+
+    if (clocktime_to_timeval(&ct, &tz, &tv)) {
+	return BLE_ATT_ERR_UNLIKELY;
+    }
+
+    /* Are we ignoring the day of week */
+    if (ble_svc_cts_current_time.day_of_week) {
+	if (timeval_to_clocktime(&tv, &tz, &ct)) {
+	    return BLE_ATT_ERR_UNLIKELY;
+	}
+	if (ct.dow != (ble_svc_cts_current_time.day_of_week % 7)) {
+	    rc = 0x80; // Data field ignored
+	}
+    }
+
+    /* Apply time change */
+    if (os_settimeofday(&tv, NULL)) {
+	return BLE_ATT_ERR_UNLIKELY;
+    }
+
+    /* Notify 
+     *  (Manual change are always notify, don't bother hinting on the offset)
+     */
+    ble_svc_cts_reference_time_updated(BLE_SVC_CTS_TIME_SOURCE_MANUAL,
+				       BLE_SVC_CTS_TIME_ACCURACY_UNKNOWN, 0);
```

Comment:
I'd say manual update is when user updates time on device, not when it is written by client. also this information should be passed from application, not automatically from service since spec does not define this anywyere. also this could trigger local time characteristic to be notified just after write which is weird, because client already knows it changed it...